File: desktopclient/src/main/java/com/goexp/galgame/gui/db/mongo/query/BrandQuery.java
package com.goexp.galgame.gui.db.mongo.query;

import com.goexp.common.db.mongo.DBQueryTemplate;
import com.goexp.common.db.mongo.ObjectCreator;
import com.goexp.galgame.common.model.BrandType;
import com.goexp.galgame.gui.model.Brand;
import org.bson.Document;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class BrandQuery {

    public final static DBQueryTemplate<Brand> tlp = new DBQueryTemplate.Builder<>("galgame", "brand", new BrandCreator())
            .build();

    private static class BrandCreator implements ObjectCreator<Brand> {

        private final Logger logger = LoggerFactory.getLogger(BrandCreator.class);

        @Override
        public Brand create(Document doc) {
            var g = new Brand();

            logger.debug("<create> doc={}", doc);

            g.id = doc.getInteger("_id");
            g.name = doc.getString("name");
            g.website = doc.getString("website");
            g.isLike = BrandType.from(doc.getInteger("type"));
            g.comp = doc.getString("comp");
            return g;
        }
    }
}

File: desktopclient/src/main/scala/com/goexp/galgame/gui/db/mongo/query/BrandQuery.scala
package com.goexp.galgame.gui.db.mongo.query

import java.util
import com.goexp.db.mongo.{DBQuery, ObjectCreator}
import com.goexp.galgame.common.Config
import com.goexp.galgame.common.db.mongo.query.CommonBrandCreator
import com.goexp.galgame.common.model.game.brand.BrandState
import com.goexp.galgame.gui.db.mongo.DB_NAME
import com.goexp.galgame.gui.db.mongo.query.StatCreators.statisticsCreator
import com.goexp.galgame.gui.model.Brand
import com.mongodb.client.model.Sorts.ascending
import com.typesafe.scalalogging.Logger
import org.bson.Document

import scala.jdk.CollectionConverters._


object BrandQuery {
  private val logger = Logger(BrandQuery.getClass)

  private val creator: ObjectCreator[Brand] = (doc: Document) => {
    logger.trace(s"<Doc> $doc")

    val parentCreator = new CommonBrandCreator(new Brand)
    val b = parentCreator.create(doc).asInstanceOf[Brand]

    b.state.set(BrandState.from(doc.getInteger("type")))
    b.tag = Option(doc.get("tag", classOf[util.List[String]])).map { l => l.asScala.toList }.orNull
    b.statistics = Option(doc.get("statistics").asInstanceOf[Document]).map(statisticsCreator.create).orNull

    logger.trace(s"<brand> ${b}")

    b

  }


  private val tpl = DBQuery[Brand](Config.DB_STRING, DB_NAME, "brand", creator)
    .defaultSort(ascending("comp"))
    .build

  def apply() = tpl

}
